sycl/source/detail/platform_impl.cpp

```diff
@@ -26,6 +27,53 @@ namespace detail {
 
 using PlatformImplPtr = std::shared_ptr<platform_impl>;
 
+// Return true if the given device 'Dev' matches with any filter
+bool matchFilterList(detail::device_filter_list *FilterList, backend Backend,
+                     info::device_type Type, unsigned int Index) {
+
+  for (const detail::device_filter &Filter : FilterList->get()) {
+    if ((Filter.Backend == Backend || Filter.Backend == backend::all) &&
+        (Filter.DeviceType == Type ||
+         Filter.DeviceType == info::device_type::all)) {
+      if (!Filter.HasDeviceNum ||
+          static_cast<unsigned int>(Filter.DeviceNum) == Index)
+        return true;
+    }
+  }
+  return false;
+}
+
+static void applyDeviceFilter(vector_class<RT::PiDevice> &PiDevices,
+                              const plugin &Plugin) {
+  detail::device_filter_list *FilterList =
+      detail::SYCLConfig<detail::SYCL_DEVICE_FILTER>::get();
+
+  // If device filter is not set use all available devices
+  if (!FilterList)
+    return;
+
+  info::device_type Type;
+  backend Backend = Plugin.getBackend();
+  unsigned int InsertIDx = 0;
+
+  for (unsigned int i = 0; i < PiDevices.size(); i++) {
+    Plugin.call<PiApiKind::piDeviceGetInfo>(
+        PiDevices[InsertIDx],
```

Comment:
Should we use `i` here instead of `InsertIDx`?

Reply:
we skip part of the devices so i points to checked device and InsertIDx to filtered devices.

Reply:
What if PiDevices = "cpu, gpu" and Filter = gpu? In this case the code do won't find gpu device because on the second iteration it will get cpu type(since InsertIDx is 0)

Reply:
You are right that it should be `I` there. But the failure you mentioned will not be reproduced because CPU and GPU devices are in different platforms and we process devices by platform. The bug has been fixed